diff --git a/modules/pam_lastlog/pam_lastlog.c b/modules/pam_lastlog/pam_lastlog.c
index 50e5a59..bd454ff 100644
--- a/modules/pam_lastlog/pam_lastlog.c
+++ b/modules/pam_lastlog/pam_lastlog.c
@@ -479,6 +479,10 @@ last_login_failed(pam_handle_t *pamh, int announce, const char *user, time_t llt
 	}
     }
 
+    if (retval != 0)
+	pam_syslog(pamh, LOG_WARNING, "corruption detected in %s", _PATH_BTMP);
+    retval = PAM_SUCCESS;
+
     if (failed) {
 	/* we want the date? */
 	if (announce & LASTLOG_DATE) {
