diff -up Linux-PAM-0.99.8.1/modules/pam_console/50-default.perms.multifunc Linux-PAM-0.99.8.1/modules/pam_console/50-default.perms
--- Linux-PAM-0.99.8.1/modules/pam_console/50-default.perms.multifunc	2007-08-22 18:45:17.000000000 +0200
+++ Linux-PAM-0.99.8.1/modules/pam_console/50-default.perms	2007-08-22 21:26:16.000000000 +0200
@@ -16,7 +16,7 @@
 
 # permission definitions
 <console>  0600 <pilot>      0660 root.uucp
-<console>  0600 <scanner>    0600 root
+<console>  0660 <scanner>    0660 root.lp
 <console>  0600 <fb>         0600 root
 <console>  0600 <kbd>        0600 root
 <console>  0600 <joystick>   0600 root
